test/integration/rack_attack_test.rb

```diff
@@ -210,6 +210,37 @@ def encode(username, password)
         assert_response :unauthorized
       end
     end
+
+    context "expontential backoff" do
+      context "with successful gem push" do
+        setup do
+          Rack::Attack::EXP_BACKOFF_LEVELS.each do |level|
+            under_backoff_limit = (Rack::Attack::EXP_BASE_REQUEST_LIMIT * level) - 1
+            push_exp_throttle_level_key = "#{Rack::Attack::PUSH_EXP_THROTTLE_KEY}/#{level}:#{@ip_address}"
+            under_backoff_limit.times { Rack::Attack.cache.count(push_exp_throttle_level_key, exp_base_limit_period**level) }
+          end
+
+          post "/api/v1/gems",
+            params: gem_file("test-0.0.0.gem").read,
+            headers: { REMOTE_ADDR: @ip_address, HTTP_AUTHORIZATION: @user.api_key, CONTENT_TYPE: "application/octet-stream" }
+        end
+
+        should "reset gem push rate limit rack attack key" do
+          Rack::Attack::EXP_BACKOFF_LEVELS.each do |level|
+            push_exp_throttle_level_key = "#{Rack::Attack::PUSH_EXP_THROTTLE_KEY}/#{level}:#{@ip_address}"
+            assert_equal 1, Rack::Attack.cache.count(push_exp_throttle_level_key, exp_base_limit_period**level)
```

Comment:
Calling `Rack::Attack.cache.count` increments the key, hence this asserts for 1, instead of 0. I couldn't get read cache to work.. something about [key prefix](https://github.com/kickstarter/rack-attack/blob/129e970d42f86c5c46978988c5c09b0eddaec35a/lib/rack/attack/cache.rb#L29).

Reply:
I'll take a look. :eyes:

Reply:
yes, you need to specify "time" part manually for read

this code works (and fails if I comment out reset in rack_attack_reset.rb)

```ruby
should "reset gem push rate limit rack attack key" do
  Rack::Attack::EXP_BACKOFF_LEVELS.each do |level|
    push_exp_throttle_level_key = "#{Rack::Attack::PUSH_EXP_THROTTLE_KEY}/#{level}:#{@ip_address}"
    period = exp_base_limit_period**level
    key = "#{Time.now.to_i / period}:#{push_exp_throttle_level_key}"

    assert_equal nil, Rack::Attack.cache.read(key)
  end
end
```

got inspired at https://github.com/kickstarter/rack-attack/blob/580368faddcb563b55ad2ec4aa4cd05aae845547/spec/rack_attack_throttle_spec.rb#L72-L78

Reply:
Thanks simi. Updated to this.
> #{Time.now.to_i / period}

Doesn't this mean if period was much smaller 1-2 seconds, key could change between setup and assert?

Reply:
It can happen with any period actually. It should be safer to wrap those specs with `freeze_time`. :+1: 

https://api.rubyonrails.org/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-freeze_time

Reply:
The idea of the key changing between the time key is set (in rack_attack.rb) and reset (in pusher.rb) did cross my mind but I assumed their `reset_count` implementation must be correct. Looking at this more closing after your comment, I don't think it is. Consider following:
```
# time during `count` in rack_attack.rb
> Time.now.to_i 
 => 1587326999
> Time.now.to_i / 100 
 => 1587326

# churn code, query pg, upload to s3. time during `reset_count`  in pusher.rb
> Time.now.to_i 
=> 1587327001
> Time.now.to_i / 100 
=> 1587327
```

To reiterate, [`key_and_expiry`](https://github.com/kickstarter/rack-attack/blob/129e970d42f86c5c46978988c5c09b0eddaec35a/lib/rack/attack/cache.rb#L62) gets called from both `count` and `reset_count` but there is no grantee that `Time.now` will be the same.

I suppose similar to their workaround for key expiry https://github.com/kickstarter/rack-attack/pull/85 we should purge be purging two keys (`rack::attack:1587326:..` and `rack::attack:1587327:..`).